test/debug/test_charptr.py

```diff
@@ -39,73 +55,117 @@ def test_charptr_use_after_implicit_arg_handle_close(compiler, python_subprocess
-    else:
-        # The garbage we override the data with will cause this error
-        assert b"UnicodeDecodeError" in result.stderr
+    content = "'use after close me!'"
+    bcontent = 'b' + content
+    for mode, msg in enumerate((content, bcontent, bcontent)):
+        if SUPPORTS_MEM_PROTECTION and False:
+            code = "mod.f({}, {}); mod.g()".format(msg, mode)
+        else:
+            code = "mod.f({0}, {1}); assert mod.g() == str({0})".format(msg, mode)
+        result = python_subprocess.run(mod, code)
+        assert result.returncode != 0
+        assert result.stdout == b""
+        if SUPPORTS_MEM_PROTECTION:
+            assert result.stderr == b""
+        else:
+            # The garbage we override the data with will cause this error
+            assert b"UnicodeDecodeError" in result.stderr
 
 
 @pytest.mark.skipif(not SUPPORTS_SYS_EXECUTABLE, reason="needs subprocess")
 def test_charptr_use_after_handle_close(compiler, python_subprocess):
     mod = compiler.compile_module("""
-        HPyDef_METH(f, "f", HPyFunc_O)
-        static HPy f_impl(HPyContext *ctx, HPy self, HPy arg)
+        HPyDef_METH(f, "f", HPyFunc_VARARGS)
+        static HPy f_impl(HPyContext *ctx, HPy self, HPy *args, HPy_ssize_t n)
         {
-            HPy arg_dup = HPy_Dup(ctx, arg);
+            if (n != 2) {
+                HPyErr_SetString(ctx, ctx->h_ValueError, "expected exactly two arguments");
+                return HPy_NULL;
+            }
+            long mode = HPyLong_AsLong(ctx, args[1]);
+            if (mode == -1)
+                return HPy_NULL;
+
+            HPy arg_dup = HPy_Dup(ctx, args[0]);
             HPy_ssize_t size;
-            const char *keep = HPyUnicode_AsUTF8AndSize(ctx, arg_dup, &size);
+            const char *keep;
+            if (mode == 0) {
+                keep = HPyUnicode_AsUTF8AndSize(ctx, arg_dup, &size);
+            } else if (mode == 1) {
+                keep = HPyBytes_AsString(ctx, arg_dup);
+            } else if (mode == 2) {
+                keep = HPyBytes_AS_STRING(ctx, arg_dup);
+            } else {
+                HPyErr_SetString(ctx, ctx->h_ValueError, "invalid mode");
+                return HPy_NULL;
+            }
             HPy_Close(ctx, arg_dup);
             return HPyUnicode_FromString(ctx, keep);
         }
 
         @EXPORT(f)
         @INIT
     """)
-    if SUPPORTS_MEM_PROTECTION:
-        code = "mod.f('use after close me!')"
-    else:
-        code = "assert mod.f('use after close me!') == 'use after close me!'"
-    result = python_subprocess.run(mod, code)
-    assert result.returncode != 0
-    assert result.stdout == b""
-    if SUPPORTS_MEM_PROTECTION:
-        assert result.stderr == b""
-    else:
-        # The garbage we override the data with will cause this error
-        assert b"UnicodeDecodeError" in result.stderr
+    content = "'use after close me!'"
+    bcontent = "b" + content
+    for mode, msg in enumerate((content, bcontent, bcontent)):
```

Comment:
Yes, that was basically the intention. Whenever I write new tests, I try to group conceptually equal tests to reduce compilation since that seems to increase test time significantly.
It would be nice if we could use `parameterize`. Maybe we can have a fixture or closure such that the module is compiled just once per session.